### PR TITLE
Simplify and optimize code around TermQueryBuilder.BinaryValues

### PR DESCRIPTION
**Pull request by original-brownbear**
Without a change in behavior, we can remove the unused ListValues as well as most of the allocations when serializing the values.
This still leaves the problem that the temporary buffer in `valueRef` could be massive in size, that will be addressed in a short follow-up that this change sets up.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 1]
```diff
@@ -787,8 +787,10 @@ public final void writeOptionalInstant(@Nullable Instant instant) throws IOExcep
         })
     );
 
+    public static final byte GENERIC_LIST_HEADER = (byte) 7;
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:790]: We were already using this impl. detail in one spot, might as well just make it a constant here for now.

File: server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java [hunk 2]
```diff
@@ -50,10 +49,9 @@
  */
 public class TermsQueryBuilder extends AbstractQueryBuilder<TermsQueryBuilder> {
     public static final String NAME = "terms";
-    private static final TransportVersion VERSION_STORE_VALUES_AS_BYTES_REFERENCE = TransportVersions.V_7_12_0;
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java:53]: Outdated, we can drop all code for pre-v7.12, in this case that means dropping the list backed `Values` and then collapsing `Values` into `BinaryValues`

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Chris!

**Pull request merged**